rangeserver/src/cache.rs

```diff
@@ -0,0 +1,73 @@
+pub mod memtabledb;
+
+use std::sync::Arc;
+
+use bytes::Bytes;
+use thiserror::Error;
+
+#[derive(Clone, Debug, Error)]
+pub enum Error {
+    #[error("Timeout Error")]
+    Timeout,
+    #[error("No space left in range cache")]
+    CacheIsFull,
+    #[error("Key not present in range cache")]
+    KeyNotFound,
+    #[error("range cache error: {0}")]
+    InternalError(Arc<dyn std::error::Error + Send + Sync>),
+}
+
+// Range cache options
+#[derive(Clone, Debug, Copy)]
+pub struct CacheOptions {
+    pub path: &'static str,   /* disk path where the db is stored */
+    pub num_write_buffers: usize,   /* num memtables */
+    pub write_buffer_size: u64, /* memtable size */
+}
+
+// callback from range cache to range server to garbage collect old epochs
+type GCCallback = fn(epoch: u64) -> ();
+pub trait Cache: Send + Sync + 'static {
+    // constructor
+    fn new(
+        cache_options: Option<&CacheOptions>,
```

Comment:
since CacheOptions is copy, no need to pass it by reference. Also I think let's not make this Option, you can define a default CacheOptions instead